src/blog/articles/node-js-race-conditions/index.md

```diff
@@ -0,0 +1,351 @@
+
+This is pretty much equivalent to the following code:
+
+```javascript
+await Promise.all([sellGrapes(), sellOlives()])
+```
+
+Using `Promise.all()` might make a little more obvious that we are scheduling some different pieces of work to run concurrently.
+
+But now that we understand the problem, how do we fix the race condition?
+
+Well, it turns out that in this simple case, we might make things right quite easily:
+
+```javascript
+async function main () {
+  await sellGrapes()
+  await sellOlives()
+  const balance = await loadBalance()
+  console.log(`Final balance: ${balance}`)
+}
+```
+
+This implementation will consistently produce the following output:
+
+```text
+sellGrapes - balance loaded: 0
+sellGrapes - balance updated: 50
+sellOlives - balance loaded: 50
+sellOlives - balance updated: 100
+Final balance: 100
+```
+
+As we can observe, `sellGrapes` is always started and completed _before_ we start `sellOlives`. This makes the two logical transactions isolated and makes sure their event won't end up being mixed together in random order.
+
+Problem solved... _vade in pacem_ dear race condition!
+
+
+## Using a mutex in Node.js
+
+OK, the previous example was illustrative, but if we are building a real game, chances are things will end up being a lot more complicated. We will probably end up having many different actions that might cause a change of balance. Those actions might be the result of a particular sequence of events and it might become hard to track down the discrete logical transactions that we have to _serialize_ in order to avoid race conditions.
+
+Ideally, we don't want to think in terms of transactions, we just need to make sure that we never read the balance if there is another event that is ready to change its value.
+
+To be able to do this we need two things:
+
+  1. Have a way to identify when _we are about to change_ the balance
+  2. Let other events wait in line until the change is completed before reading the balance
+
+We could say that when _we are about to change_ the balance we enter a _critical path_ and that we don't want to intermingle events from different logical transactions in a critical path.
+
+One way to achieve this is by using a *Mutex* (which stands for [**mut**ual **ex**clusion](https://en.wikipedia.org/wiki/Mutual_exclusion)).
+
+A mutex is a mechanism that allows synchronising access to a shared resource.
+
+We can see a mutex as a shared object that allows us to mark when the code execution is entering and exiting from a critical path. In addition to that, a mutex can help us to queue other logical transactions that want to access the same critical path while one transaction is being processed.
+
+
+## Using `async-mutex`
```

Comment:
You can also synchronize using just promises. Just in case you haven't considered it.

```
let mutex = Promise.resolve()


function op() {
  const res = mutex.then(() => 
   critical path
 )
  // so it never throws
  mutex = res.catch(err => {})
  return res
}
```

Reply:
what if you have multiple concurrent call to `op()`? wouldnt' they get all resolved after the current unit of work on the critical path is completed?

Reply:
Multiple calls will be queued one after the other, because `mutex` will be replaced (synchronously) with the promise representing the current critical path, which in turn will be executed only when the previous critical path completes. It would be like having:

```javascript
mutex.then(() => 
   critical path for Op 1
 )
.catch(err => {})
.then(() => 
   critical path for Op 2
).catch(err => {})
.then(() => 
   critical path for Op 3
 )
```

Reply:
right, `mutex` is a global instance. makes total sense :)  I might add an additional section to illustrate this idea before the ending of the article.

Reply:
Added